IronFoundry.Warden/Tasks/ProcessCommand.cs

```diff
@@ -66,6 +94,89 @@ protected TaskCommandResult RunProcess(string workingDirectory, string executabl
             return new TaskCommandResult(exitCode, null, null);
         }
 
+        /// <summary>
+        /// Create a log in the container holding the id of the process.
+        /// </summary>
+        /// <returns></returns>
+        private TempFile CreatePidLog(int pid)
+        {
+            var tempFile = container.FileInContainer(PidLogRelativePath);
+
+            System.IO.File.WriteAllText(tempFile.FullName, pid.ToString(), Encoding.ASCII);
+            
+            return tempFile;
+        }
+
+        /// <summary>
+        /// Create a log in the container with the environment variables used to start the process.
+        /// </summary>
+        private TempFile CreateProcessEnvLog(IDictionary<string, string> processEnv)
+        {
+            var tempFile = container.FileInContainer(EnvLogRelativePath);
+            
+            var envLogBuilder = new StringBuilder();
+            foreach (var kv in processEnv)
+            {
+                string line = string.Format("{0} = {1}", kv.Key, kv.Value);
+                envLogBuilder.AppendLine(line);
+            }
+
+            System.IO.File.WriteAllText(tempFile.FullName, envLogBuilder.ToString(), Encoding.UTF8);
+
+            return tempFile;
+        }
+
+        /// <summary>
+        /// Prepare the complete environment block that will be used based on the environment
+        /// variables that were specified in the request.
+        /// </summary>
+        /// <remarks>
+        /// When starting a process, you can let the process load the default environment or 
+        /// you can specify the exact variables to use by setting ProcessStartInfo.EnvironmentVariables.
+        /// 
+        /// If you specify the value of any variable, then you have to specify all of them, no defaults
+        /// will be provided
+        /// 
+        /// This method will generate the default environment variables to use and then upsert the ones
+        /// specified.  
+        /// </remarks>
+        /// <remarks>
+        /// For the variable values specified, any @ROOT@ prefixes found, will be replaced
+        /// with the full path to the container root.
+        /// </remarks>
+        private IDictionary<string, string> PrepareEnvironment(IDictionary<string, string> env)
+        {
+            if (env.IsNullOrEmpty())
+            {
+                return new Dictionary<string, string>(0);
+            }
+
+            // Root the provided environment values
+            var rootedEnv = env.ToDictionary(kv => kv.Key, kv => container.ConvertToPathWithin(kv.Value));
+
+            // Generate a new environment containing all the default values.
+            EnvironmentBlock defaultEnv = EnvironmentBlock.GenerateDefault();
+            
+            // Generate an environment with the machine variables
+            var machineVars = Environment.GetEnvironmentVariables(EnvironmentVariableTarget.Machine);
```

Comment:
It looks like the default environment block already contains the 'machine' level entries.  If so, there's no need to merge in the machine-specific vars explicitly.